weighted_graphs.py

```diff
@@ -0,0 +1,214 @@
+class WeightedGraphs():
+  def dijkstras_algorithm(self, graph, starting_vertex):
+    """
+      my implementation and understanding is heavily taken from bradfield's
+      article https://bradfieldcs.com/algos/graphs/dijkstras-algorithm/ and
+      Bari's video https://www.youtube.com/watch?v=XB4MIexjvY0
+
+      I've commented almost every line of the function and printed a whole
+      bunch of info so that it is much more obvious to me how the algorithm
+      runs, especially the part about updating the dict of costs
+
+      Note that I put 'cost' rather than 'distance' to make this more generic
+      because at the end of the day, a weighted graph just says that oh, it
+      takes X to go from Y to Z. and that X can be distance, time, etc.
+      at the end of the day, X is just a cost.
+    """
+
+    import pprint
+    pprint.pprint(graph)
+    # let's say (source_vertex, target_vertex, distance) and we have
+    # [(a, b, 5), (a, c, 7), (c, d, 3)]
+
+    # cost_to_starting_vertex == costs_to_start_vert
+    # initialise the dict.
+    # for vertex in graph, set the distance from the starting vertex to infinity
+    costs_to_start_vert = {vertex: float('infinity') for vertex in graph}
+
+    # obviously, the distance from the starting vertex to itself is 0. so update that
+    costs_to_start_vert[starting_vertex] = 0
+
+    # dist_frm_start_vert == {'a': 0, 'b': inf, 'c': inf, 'd': inf}
+
+    import heapq
+    # priority queue time
+
+    # we process vertexes that have the lowest weight to starting_vertex,
+    # and that are directly connected to starting_vertex
+    # process vertex `a` first.
+    # `b` and `c` are directly connected to `a`, so process them next
+    # `b` is weight 5 to `a`. `c` is weight 7. so process vertex `b` first
+    # and so on.
+
+    # let's go. initialise the priority queue with (0, starting_vertex)
+    # makes sense cos hey, starting_vertex is a weight of 0 to itself
+    priority_queue = [(0, starting_vertex)]
+
+    # while the queue isn't empty...
+    while priority_queue:
+      # pop the top of the heap off. this is the vertex with the lowest weight to the starting_vertex
+      curr_cost_to_start_vert, curr_vert = heapq.heappop(priority_queue)
+      print(f"**curr_vert {curr_vert}**")
+      print(f"existing_cost_to_{starting_vertex}: {costs_to_start_vert[curr_vert]}. costs_to_{starting_vertex}: {costs_to_start_vert}")
+
+      # note says "nodes can get added to the priority queue many times
+      # only process a vertex the first time we remove it from the priority queue"
+      # ^ not sure what that means and why it's important we do this
```

Comment:
Consider a node that has multiple incoming edges.  Each time you visit that edge, you might add the node to the priority queue again.

"Why not just add it once?"
Because you don't know which edge the shortest path will need.

```
A---C
 \ /
  B
```

Suppose A -> C has weight 100, but A -> B and B -> C both have weight 1.